exonum/src/events/mod.rs

```diff
@@ -12,377 +12,182 @@
-        self.event_loop.run(&mut self.inner)
-    }
-    fn run_once(&mut self, timeout: Option<usize>) -> ::std::io::Result<()> {
-        self.event_loop.run_once(&mut self.inner, timeout)
-    }
-    fn get_time(&self) -> SystemTime {
-        SystemTime::now()
-    }
-    fn channel(&self) -> MioChannel<H::ApplicationEvent, H::Timeout> {
-        MioChannel {
-            inner: self.event_loop.channel(),
-            address: *self.inner.network.address(),
-        }
-    }
-}
-
-trait LogError {
-    fn log_error<S: AsRef<str>>(self, msg: S);
+#[derive(Debug)]
+pub struct EventsAggregator<S1, S2, S3>
+where
+    S1: Stream,
+    S2: Stream,
+    S3: Stream,
+{
+    done: bool,
+    timeout: S1,
+    network: S2,
+    api: S3,
 }
 
-impl<E> LogError for result::Result<(), E>
+impl<S1, S2, S3> EventsAggregator<S1, S2, S3>
 where
-    E: fmt::Display,
+    S1: Stream,
+    S2: Stream,
+    S3: Stream,
 {
-    fn log_error<S: AsRef<str>>(self, msg: S) {
-        if let Err(error) = self {
-            error!("{}, an error occurred: {}", msg.as_ref(), error);
+    pub fn new(timeout: S1, network: S2, api: S3) -> EventsAggregator<S1, S2, S3> {
+        EventsAggregator {
+            done: false,
+            network: network,
+            timeout: timeout,
+            api: api,
         }
     }
 }
 
-#[derive(Debug)]
-pub struct Error {
-    message: String,
-}
-
-pub type Result<T> = result::Result<T, Error>;
-
-impl Error {
-    pub fn new<T: Into<String>>(message: T) -> Error {
-        Error { message: message.into() }
-    }
-}
-
-impl fmt::Display for Error {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "Events error: {}", self.message)
-    }
-}
+impl<S1, S2, S3> Stream for EventsAggregator<S1, S2, S3>
+where
+    S1: Stream<Item = NodeTimeout>,
+    S2: Stream<
+        Item = NetworkEvent,
+        Error = S1::Error,
+    >,
+    S3: Stream<
+        Item = ExternalMessage,
+        Error = S1::Error,
+    >,
+{
+    type Item = Event;
+    type Error = S1::Error;
 
-impl error::Error for Error {
-    fn description(&self) -> &str {
-        &self.message
+    fn poll(&mut self) -> Poll<Option<Event>, Self::Error> {
+        if self.done {
+            Ok(Async::Ready(None))
+        } else {
+            // Check timeout events
+            match self.timeout.poll()? {
+                Async::Ready(Some(item)) => {
+                    return Ok(Async::Ready(Some(Event::Timeout(item))));
+                }
+                // Just finish stream
+                Async::Ready(None) => {
+                    self.done = true;
+                    return Ok(Async::Ready(None));
+                }
+                Async::NotReady => {}
+            };
+            // Check network events
+            match self.network.poll()? {
+                Async::Ready(Some(item)) => {
+                    return Ok(Async::Ready(Some(Event::Network(item))));
+                }
+                // Just finish stream
```

Comment:
Are such comments useful? There are three identical comments here, perhaps they can be replaced by the single comment (with a better explanation) at the beginning of the function.